components/browser/session-storage/src/main/java/mozilla/components/browser/session/storage/SessionStorage.kt

```diff
@@ -40,12 +41,14 @@ class SessionStorage(
 
     /**
      * Reads the saved state from disk. Returns null if no state was found on disk or if reading the file failed.
+     *
+     * @param predicate an optional predicate applied to each tab to determine if it should be restored.
      */
     @WorkerThread
-    fun restore(): RecoverableBrowserState? {
+    fun restore(predicate: (RecoverableTab) -> Boolean = { true }): RecoverableBrowserState? {
         synchronized(sessionFileLock) {
             val file = getFileForEngine(context, engine)
-            return stateReader.read(engine, file)
+            return stateReader.read(engine, file, predicate)
```

Comment:
Interesting, you went a different route. I think I'd rewrite the `RecoverableBrowserState` (`copy()`) here if a filter was provided and leave everything else as-is and avoid passing this all the way down? 🤔

Reply:
Yes, to keep it all in one place. In `JsonReader.browsingSession` we already process the `selectedTabId` which we would have to do there and here again also in case the selected tab was filtered out. We'd also put the tabs in and then remove them again if the `predicate`returns false. This way we filter right away and not even add....